### ISSUE TITLE
Bug: Reshard DMA reads OOB source tiles when K-dimension is not divisible by destination grid column count

### ISSUE THREAD

**Issue by hkwonTT** [tenstorrent/tt-mlir#8815]
## Summary

When resharding a tensor from an N-wide core-column layout to an M-wide layout where `K_tiles % M != 0`, the grid-aware alignment pads K to `ceil(K_tiles/M)*M`. The reshard DMA then issues NoC reads for the phantom padding tiles, which map to source column N — out of bounds for a source that only spans columns 0..N-1.

## Root Cause

The reshard DMA computes how many K-tiles each destination column needs based on the padded (aligned) K size, not the actual logical K size. There is no guard to prevent reading phantom tiles beyond the source tensor boundary.

Concrete numbers:

| Parameter | Value |
|---|---|
| Activation shape | `[256, 4096]`, K = 128 tiles |
| Source layout | 8x8 grid -> 16 K-tiles per column |
| Destination layout | 8x7 grid -> K padded to `ceil(128/7)*7 = 133` tiles, 19 tiles/column |
| OOB read | tiles 128-132 mapped to **source column 8** (valid range: 0-7) |

The 8x7 destination grid in the reproducer is created via `d2m.spatial` restricting compute to 7 columns, but the reshard bug itself is independent of `d2m.spatial` — any code path that produces an (N-wide source, M-wide destination) reshard with `K_tiles % M != 0` will trigger it.

## Observed Failure Modes

**Case 1 — standalone n150 (FABRIC_DISABLED)**
ETH core is dormant; the phantom read returns garbage immediately. `submit()` returns but the matmul reduction has accumulated phantom values -> **wrong output (PCC << 0.99)**.

**Case 2 — multi-chip host / LLMBox (FABRIC_1D_RING, or UMD opens multiple chips)**
ETH cores are in forwarding mode. The phantom NoC read packet is forwarded around the ring and never acknowledged -> **`async_read_barrier` never returns -> infinite device hang**.

> Note: The D2MBuilder Python API (`get_metal_tensor_layout`) happens to block this layout via `assert l % g == 0`, so the bug is only reachable through the compiler pipeline directly. The reproducer therefore uses the MLIR file directly.

## Reproducer

Branch: `hkwon/reshard_padding_hang`

Two files:

- **MLIR** (`test/ttmlir/Silicon/TTMetal/n150/spatial/reshard_padding_hang_repro.mlir`): defines the buggy D2M IR with explicit padded layouts (`dim_alignments=32x224`).
- **Python** (`test/python/golden/d2m/test_reshard_padding_bug.py`): compiles the MLIR via `ttmlir-opt` + `ttmlir-translate` and executes the flatbuffer through `_ttmlir_runtime`.

```bash
# Case 1 (standalone n150 -- wrong output):
pytest test/python/golden/d2m/test_reshard_padding_bug.py \
    ::test_reshard_padding_wrong_output \
    --sys-desc=ttrt-artifacts/system_desc.ttsys

# Case 2 (multi-chip host -- WARNING: hangs device, requires manual reset):
pytest test/python/golden/d2m/test_reshard_padding_bug.py \
    ::test_reshard_padding_hang \
    --sys-desc=ttrt-artifacts/system_desc.ttsys
```

```python
# test/python/golden/d2m/test_reshard_padding_bug.py  (abbreviated)

_MLIR_SRC = (
    Path(__file__).parents[3]
    / "ttmlir/Silicon/TTMetal/n150/spatial/reshard_padding_hang_repro.mlir"
)

@pytest.mark.parametrize("target", ["ttmetal"])
@pytest.mark.parametrize("fabric_config", [tt_runtime.runtime.FabricConfig.DISABLED])
def test_reshard_padding_wrong_output(target, fabric_config, request, device):
    """Case 1: FABRIC_DISABLED, standalone n150 -> PCC < 0.99 (wrong output)."""
    fbb = _compile_to_flatbuffer(request.config.getoption("--sys-desc"))
    dev_out = _run_and_get_output(device, fbb)
    ref_out = _reference_output()
    pcc = torch.corrcoef(torch.stack([ref_out.flatten(), dev_out.flatten()]))[0, 1].item()
    assert pcc > 0.99, f"PCC {pcc:.6f} < 0.99: phantom tile accumulation in matmul"


@pytest.mark.parametrize("target", ["ttmetal"])
@pytest.mark.parametrize("fabric_config", [tt_runtime.runtime.FabricConfig.FABRIC_1D_RING])
def test_reshard_padding_hang(target, fabric_config, request, device):
    """Case 2: FABRIC_1D_RING -> device hang (WARNING: requires manual device reset)."""
    fbb = _compile_to_flatbuffer(request.config.getoption("--sys-desc"))
    dev_out = _run_and_get_output(device, fbb)   # hangs here
    ...
```

## Expected Behavior After Fix

- `test_reshard_padding_wrong_output` passes (PCC >= 0.99): the reshard DMA no longer reads phantom tiles, and the matmul output is numerically correct.
- `test_reshard_padding_hang` completes without hanging: `submit()` returns normally and the device does not need to be reset.
